AWSIoT/Internal/SocketRocket/AWSSRWebSocket.m

```diff
@@ -1230,6 +1232,40 @@ - (void)_addConsumerWithScanner:(stream_scanner)consumer callback:(data_callback
     [self _pumpScanner];
 }
 
+- (void)_scheduleCleanup
+{
+    @synchronized(self) {
+        if (_cleanupScheduled) {
+            return;
+        }
+
+        _cleanupScheduled = YES;
+
+        // Cleanup NSStream delegate's in the same RunLoop used by the streams themselves:
+        // This way we'll prevent race conditions between handleEvent and SRWebsocket's dealloc
+        NSTimer *timer = [NSTimer timerWithTimeInterval:(0.0f) target:self selector:@selector(_cleanupSelfReference:) userInfo:nil repeats:NO];
+        [[NSRunLoop AWSSR_networkRunLoop] addTimer:timer forMode:NSDefaultRunLoopMode];
+    }
+}
+
+- (void)_cleanupSelfReference:(NSTimer *)timer
+{
+    @synchronized(self) {
+        // Nuke NSStream delegate's
+        _inputStream.delegate = nil;
+        _outputStream.delegate = nil;
+
+        // Remove the streams, right now, from the networkRunLoop
+        [_inputStream close];
+        [_outputStream close];
+    }
```

Comment:
As noted in the review comment, I think we need to call `[timer invalidate]` in this `@synchronized` block

Reply:
Great call out.  I read through the documentation here:
https://developer.apple.com/documentation/foundation/nstimer

> Once scheduled on a run loop, the timer fires at the specified interval until it is invalidated. A nonrepeating timer invalidates itself immediately after it fires. However, for a repeating timer, you must invalidate the timer object yourself by calling its invalidate method.


Since this is a non-repeating timer, I believe that calling `invalidate` is not needed, because we are passing in `.. repeats:NO`:
```
 NSTimer *timer = [NSTimer timerWithTimeInterval:(0.0f) target:self selector:@selector(_cleanupSelfReference:) userInfo:nil repeats:NO];
```

Reply:
Thanks, you're absolutely right.